apps/src/codegen.js

```diff
@@ -52,23 +52,29 @@ exports.evalWith = function (code, options, legacy) {
  * Generate code for each of the given events, and evaluate it using the
  * provided APIs as context.
  * @param apis Context to be set as globals in the interpreted runtime.
- * @param events Mapping of hook names to the corresponding blockly block name.
- * @param generator Function for generating code for a given block name.
+ * @param events Mapping of hook names to the corresponding handler code.
+ * @param evalCode Optional extra code to evaluate.
  * @return {{}} Mapping of hook names to the corresponding event handler.
  */
-exports.evalWithEvents = function (apis, events, generator) {
+exports.evalWithEvents = function (apis, events, evalCode = '') {
   let interpreter, currentCallback;
-  let code = '';
   const hooks = {};
   Object.keys(events).forEach(event => {
-    hooks[event] = () => {
-      currentCallback(event);
+    hooks[event] = (...args) => {
+      const eventArgs = {name: event, args: args};
+      currentCallback(exports.marshalNativeToInterpreter(interpreter, eventArgs, null, 5));
       interpreter.run();
     };
-    code += `function ${event}(){${generator(events[event])}};`;
+    const {code, args} = events[event];
+    evalCode += `this['${event}']=function(${args ? args.join() : ''}){${code}};`;
```

Comment:
I think args will always be truthy here (and an empty array in the case where no args are provided)